[fixed-hash] Major overhaul of `construct_hash!` macro

Robbepop opened the pull request:
# Overhaul of the entire sub-crate `fixed-hash`

## Roadmap

### Pre-merging

- Sort out the last few nits with new features
    - [x] Finalize documentation
    - [x] Remove deprecated features
    - [x] Test new features
    - [x] Clean up public API

- Merge this PR after
    - [x] Merge of [PR](https://github.com/paritytech/parity-common/pull/72)
    - [x] After release of [Rust 1.30](https://github.com/rust-lang/rust/blob/stable/RELEASES.md)

- [x] Bump version to `0.3-beta.0` and publish on crates.io.

### Post-merging

Transit all below dependencies to the `0.3.0-beta.X` version of `fixed-hash`. Note that the `pwasm-*` crates currently depend on `parity-hash` instead of `fixed-hash` version `0.2`.

- `pwasm-std` will provide new `uint` instances and updated `fixed-hash` instances without conversion between `uint` and `fixed-hash` since I cannot see that it has been provided yet
- `pwasm-ethereum` will no longer link to `parity-hash` and instead use `pwasm-std`
- `pwasm-abi` same as `pwasm-ethereum`
- `ethbloom` simply update `fixed-hash`
- `ethereum-types` update `fixed-hash`, but already provides `uint` conversions itself, so no need for providing them by `fixed-hash` either
- `substrate-primitives` update `fixed-hash` but does not need `uint` conversion

- Make any changes that might come up during the work on step 4 and publish `0.3-beta.X` as needed
- Bump version to `0.3`, create a tag `fixed-hash-0.3` and publish `0.3`

## Open Questions

- What to do with `impl_hash_uint_conversions` macro?
    - The problem: Why should it be defined in `fixed-hash` and not in `uint`?
    - Possible solution: Maybe add another crate to combine `fixed-hash` and `uint` and allow for this macro.

## Summary of changes

- Major overhaul of the `construct_hash` macro
    - Possibility to declare types with visibility modifier (not only `pub`)
    - Possibility to add custom doc comment
    - Syntactically more conforming to default Rust syntax
    - Updated documentation and added examples
- Major overhaul of the `impl_hash_conversion` macro
    - No longer requires sizes of input types
    - Added proper error handling for some conversions
    - Updated documentation and added examples
- The `std` feature is now a default feature for this crate
- Removed items that have been deprecated in `0.2.5`
- Removed bad usage of `AsRef<Self>` impl
- Removed `serialize` crate feature

Robbepop commented:
As far as I can tell the build failures reported by TravisCI and Appveyor are **not** related to `fixed-hash` and the changes that have been done here.
To verify that yourself you can try to test this crate in isolation locally.

Robbepop commented:
@tomusdrw Thank you for the feedback!

I have found some other things in the crate that could be improved.
Also the exported macro `impl_hash_conversions` could have a simpler interface and using `path` matchers instead of `ident` in its signature.

I didn't fix it since this PR is already pretty big but in the end it should be fixed as well before we release another major version of the `fixed-hash` crate.

Robbepop commented:
TravisCI and Appveyor are showing errors because of [this](https://github.com/rust-lang/rust/issues/55241) issues which seems to be unrelated to the work in this PR.

dvdplm commented on fixed-hash/src/hash.rs:
```diff
@@ -177,197 +211,234 @@ macro_rules! construct_hash {
+	($large_ty:ident, $small_ty:ident) => {
+		impl From<$small_ty> for $large_ty {
+			fn from(value: $small_ty) -> $large_ty {
+				use std::mem;
```
Something fishy is going on with this macro; on this branch I get warnings about missing std. If the conversions need to work in both `std`/`no_std` we should use `core::mem` (and sort out the implications in `lib.rs`). I also get a bunch of `type mismatch` test failures when enabling this feature.
In the PR description you mentioned "testing in isolation". Not sure what you meant by that exactly but I copied the files out of `parity-common` and fixed up `Cargo.toml` and ran the tests: same failures. Can you elaborate further on this aspect?

Robbepop replied:
Unfortunately I can't find where I have written something about "testing in isolation" and can't remember either. :(
I hope the GitHub downtime is over soon and things are syncing again correctly.

dvdplm replied:
> To verify that yourself you can try to test this crate in isolation locally.

Robbepop replied:
@pepyakin and me were trying to find a solution together.

The solution is to include the `"std"` feature in the `default` features.
This means that you have to build the crate with the cargo flag `--no-default-features` for a proper `no_std` environment. Fortunately this fixes the problems with doc tests above.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -177,197 +211,234 @@ macro_rules! construct_hash {
-/// last bytes, e.g. `From<H256> for H160` uses bytes 12..32
-/// CAUTION: make sure to call with correct sizes and the bigger type first or
-/// bad things will happen!
+/// Implements lossful conversions between the given types.
```
thanks, will correct that! :)

Robbepop replied:
I have fixed it but GitHub doesn't show "lossy". Cannot tell why. Maybe still due to the GitHub downtime?

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -16,96 +20,124 @@ pub fn clean_0x(s: &str) -> &str {
             }
 
+			/// Returns a mutable raw pointer to the value.
+			#[inline]
```
Wow those indentations are among the hardest to fix ...

Robbepop replied:
I really can't fix it and don't know why. lol

Robbepop replied:
Can you still merge the PR? I will try to fix this on GitHub in another PR.

dvdplm replied:
Yeah, I see this is fixed when pulling. Not blocking merge.

Robbepop replied:
So we can resolve this discussion I guess. :)

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -579,13 +650,36 @@ macro_rules! impl_quickcheck_arbitrary_for_hash {
+		///
+		/// # Note
+		///
+		/// Mainly used for addresses in ethereum and solidity context.
```
Same as above: GitHub shows old version, not the updated one. Cannot tell why.

Robbepop commented:
> Much better docs&examples, and I like the type visibility changes (although: nightly only is a no-no for this to be merged). Tests are failing and I'm not sure exactly why.
The feature set listed in Cargo.toml looks odd (not this PRs fault, but worth fixing if we're introducing breaking changes) and plain wrong (e.g. we don't have any serde code).
Past experience tells me new issues tend to pop up we start using the parity-common crates in parity-ethereum or wherever they're used; it'd be fantastic if you could start a PR in the projects that use fixed-hash to make sure the changes work well there. This crate in particular should work as a replacement for parity-hash so it'd be good to check what needs to be done for that to happen as well.

The features set really looks a bit odd.
I will address this as soon as possible but we really should come to an end concerning this PR.

The PR for other projects to switch to `fixed-hash` from `parity-hash` (or something else) is actually a good idea and already partially in the pipeline. ;)

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -16,96 +20,124 @@ pub fn clean_0x(s: &str) -> &str {
 #[macro_export]
 macro_rules! construct_hash {
-	($from: ident, $size: expr $(, $m:meta)*) => {
+	( $(#[$attr:meta])* $visibility:vis struct $name:ident ( $n_bytes:expr ); ) => {
```
We both were right!
This feature is going to be stable with Rust 1.30 that is going to be released in 3 days (2018-10-25).
https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1300-2018-10-25

dvdplm commented on fixed-hash/Cargo.toml:
```diff
@@ -20,7 +20,7 @@ serde_derive = { version = "1.0", optional = true }
 
 [features]
-default = ["libc"]
+default = ["std", "libc"]
```
The `serialize` feature should be removed too I think.

Robbepop replied:
Could you elaborate on that? As far as I see it may be useful when depending crates are using `serde`.

Robbepop replied:
True, and since you can simply annotate the hash types in `construct_hash!` macro ínvokation you really do not need serde support. You are absolutely right and I am going to remove it. :)

dvdplm commented on fixed-hash/src/hash.rs:
```diff
@@ -6,7 +6,11 @@
+/// Return the given string `s` without the `0x` at the beginning of it, if any.
+#[deprecated(
+	since = "0.3.0",
+	note = "the api did not fit into this library, implement it yourself if needed"
```
I think we can release a 0.2.5 with the deprecation notice and remove it in 0.3. It makes very little sense for it to be here.

fckt replied:
I support both reasoning, but I like the idea to put rationals as a general rule, even if it seems redundant in some cases (b/c humans are weaker than machines..)

dvdplm commented on fixed-hash/src/hash.rs:
```diff
@@ -418,24 +500,28 @@ macro_rules! impl_heapsize_for_hash {
 		impl $from {
 			/// Get a hex representation.
-			#[deprecated(note="Use LowerHex or Debug formatting instead.")]
+			#[deprecated(note = "Use LowerHex or Debug formatting instead.")]
```
This has been deprecated for a long time, let's remove it.

Robbepop replied:
good idea! :)

Robbepop commented:
@dvdplm I think it is best to release a `0.3` version since there are many breaking changes already. Then we remove all the deprecated items and release another `0.4` version. So people can use `0.3` to transit more easily to `0.4`.

dvdplm commented:
> I think it is best to release a `0.3` version since there are many breaking changes already.

Yes. My suggestion was: keep this PR as the basis for `0.3` and add a new PR for a version `0.2.5` that just adds the deprecation notices. Then we release both `0.2.5` and `0.3`. What do you think?

Robbepop commented:
> > I think it is best to release a `0.3` version since there are many breaking changes already.
> 
> Yes. My suggestion was: keep this PR as the basis for `0.3` and add a new PR for a version `0.2.5` that just adds the deprecation notices. Then we release both `0.2.5` and `0.3`. What do you think?

I think it is better to provide replacement API together with deprecations. So a patch that only includes deprecations without all the new stuff that is replacing it is not useful I think. But of course we could do that `0.2.5` later if it is wanted or needed.

dvdplm commented:
> So a patch that only includes deprecations without all the new stuff that is replacing it is not useful I think.

I disagree, but it's not important. Let's proceed like you suggest. :)

dvdplm commented:
At this point I think the road ahead looks something like this:

1. sort out the last few nits with features
2. fix the failing tests
3. bump version to `0.3-beta.0` and publish
4. start PRs in `parity-ethereum`, `substrate` and `pwasm-*` to use the beta version
5. make any changes to `fixed-hash` that might come up during the work on step 4; publish `0.3-beta.x` as needed. (Hopefully this step is not needed at all)
6. bump version to `0.3`; merge the PR; tag repo with `fixed-hash-0.3`; publish `0.3`

Robbepop commented:
> > So a patch that only includes deprecations without all the new stuff that is replacing it is not useful I think.
> 
> I disagree, but it's not important. Let's proceed like you suggest. :)

I thought again about your idea and have to admit that it might be better than what I have had in mind. So let's do the `0.2.5` version for the deprecations and then release the cleaned up `0.3.0` version afterwards.

Robbepop commented:
Made a PR for just the deprecations and the version bump to version `0.2.5`.

dvdplm commented:
re-test failures. Beyond the compiler bug this is what I see locally:

cargo t => OK
cargo t --features=std => OK
cargo t --features=heapsizeof => OK
cargo t --features=impl_quickcheck_arbitrary => OK
cargo t --features=uint => OK
cargo t --features=serialize => FAIL (doc test I think)
cargo t --features=uint_conversions => FAIL (6 errors)

Travis runs `cargo test --features=std,heapsizeof,uint_conversions` which fails with the same 6 errors, as do `cargo t --all-features`.

fckt commented on fixed-hash/src/hash.rs:
```diff
@@ -16,157 +20,224 @@ pub fn clean_0x(s: &str) -> &str {
 			#[inline]
-			fn deref_mut(&mut self) -> &mut [u8] {
-				&mut self.0
+			pub fn zero() -> $name {
```
`$from` -> `$name` makes a lot of sense! The `$from` variable is veeery confusing, indeed..

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -16,157 +20,224 @@ pub fn clean_0x(s: &str) -> &str {
-
+		impl From<$name> for [u8; $n_bytes] {
 			#[inline]
-			fn deref(&self) -> &[u8] {
```
According to the Rust documentation and API style guidelines `Deref` and `DerefMut` implementations are meant exclusively for smart pointer types.

> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and DerefMut were designed specifically to accommodate smart pointers. Because of this, Deref should only be implemented for smart pointers to avoid confusion.

Robbepop replied:
In version 0.3 we still have `as_ref` and the new `as_bytes` methods to replace all uses of `Deref` and `DerefMut`. :)

fckt replied:
makes sense

Robbepop replied:
resolved then

pepyakin commented on fixed-hash/src/hash.rs:
```diff
@@ -16,157 +20,224 @@ pub fn clean_0x(s: &str) -> &str {
-		}
 
-		impl ::core::ops::DerefMut for $from {
+			/// Synonym for `new()`. Prefer to new as it's more readable.
```
`Prefer to new`? A grammar seems a bit off

Robbepop replied:
This documentation line is not from this PR. Don't know why it is shown as that.

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+            H64::from([
+                0x01, 0x23, 0x45, 0x67,
+                0x89, 0xAB, 0xCD, 0xEF
+            ])
```
spaces

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+	#[test]
+	fn empty_str() {
+		use core::str::FromStr;
+        assert!(H64::from_str("").is_err())
```
spaces

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+	#[test]
+	fn invalid_digits() {
+		use core::str::FromStr;
+        assert!(H64::from_str("Hello, World!").is_err())
```
spaces

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+    #[test]
+    fn too_many_digits() {
+		use core::str::FromStr;
+        assert!(H64::from_str("0123456789ABCDEF0").is_err())
```
spaces

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+		use core::str::FromStr;
+
+        assert_eq!(
+            H64::from_str("0123456789ABCDEF").unwrap(),
```
What will it do if given less than 8 bytes? In any case, I think a test for that case won't hurt.

pepyakin commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,327 @@
+	fn valid() {
+		use core::str::FromStr;
+
+        assert_eq!(
```
spaces

Robbepop commented:
Thank you all for your reviews!

I think I am done for everything that needed to be done for an initial `0.3-beta.0` release and will merge when you are all okay with all the recent changes.

- Completely written a new unit test suite.
- Updated all docs.
- Some new APIs and removed others (read commit log for more infos)

dvdplm commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
 
-/// Implements conversion to and from a hash type and the equally sized unsigned int.
-/// CAUTION: Bad things will happen if the two types are not of the same size!
-#[cfg(feature="uint_conversions")]
```
What is the plan for these? Move them to `uint`? Own crate? Just remove them? They are currently in use in `ethereum-types` so we need some kind of upgrade path for consumers.

Robbepop replied:
I had the idea to move them to `pwasm-std` for the `pwasm-*` crates and also provide `uint` types by `pwasm-std` so it fits again for it. And for `ethereum-types` I would do the same I guess. Maybe just another crate to unify both? It actually is not feasible to make `uint` depending on `fixed-hash` since it is built up from `u64` words instead of bytes and would require another major overhaul. However, `uint` is far more complex than `fixed-hash`.

So to summarize: The best way is to have this functionality in the crate that actually needs this interoperability **or** in another small utility crate featuring both types.

dvdplm replied:
Hmm, a tricky one! My personal opinion is that it's probably less disruptive/more pragmatic to just keep these conversions in `fixed-hash` for the time being and refactor later – maybe we can make the consuming code use the conversions less or not at all and we can just remove this? If you feel strongly this is not a good path ahead, I can be swayed though (though we can't merge before we have a solution to this as it'd make upgrading impossible).

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
-				}
+		#[cfg(all(
+			feature = "heapsize-support",
+			feature = "libc",
```
I will look into it.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
-				ret
+		/// Utilities using the `rand` crate.
+		#[cfg(feature = "rand-support")]
+		impl $name {
```
I can understand your critics about expanding the API. I just did the same as I have done already in my crate `apint`. There the `rand`-based API is a good addition and allows for easy testing with the `*_using` APIs and allows for user friendly usage by the APIs with the default RNG. :)

Robbepop replied:
Keep in mind that the `rand` API existed earlier but wasn't testable since it was missing the `*_using` methods. So basically the new API allows for greater control and better testability.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
-		impl Clone for $from {
-			fn clone(&self) -> $from {
-				let mut ret = $from::new();
+		#[cfg_attr(feature = "dev", allow(expl_impl_clone_on_copy))]
```
This is from an old revision and to be honest I thought that the `dev` feature is something magical. But ofc we can remove it if it doesn't exist at all.

dvdplm replied:
@debris do you know the origin of `features="dev"` per chance?

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
 		}
 
-		impl Copy for $from {}
+		impl $crate::core::marker::Copy for $name {}
```
Maybe we could even do this with the simple addition of attributes via `#[derive(Copy)]` on the user side. This would be pretty cool. :)

Robbepop replied:
According to the `Copy` reference we should keep the behaviour of implementing `Copy` by default for all hash types: https://doc.rust-lang.org/std/marker/trait.Copy.html#when-should-my-type-be-copy

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
-serialize = ["serde", "serde_derive"]
-uint_conversions = ["uint"]
+default = ["std", "libc", "rand-support", "rustc-hex-support", "byteorder-support"]
+std = ["rustc-hex/std", "rand/std", "byteorder/std"]
```
So you say that all of those features should be disabled by default? I can do that but in my opinion it is always a good thing to have the common uses active by default. This is for example `byteorder-support` and I would even say `rustc-hex-support`.

dvdplm replied:
No, that's not what I meant. I think that by having `rand/std` you tell cargo to "use the rand feature and use it with the std feature of rand enabled", i.e. `default = ["std"]` implies that `rustc-hex`, `rand` and `byteorder` will all be enabled (and will all be configured to use their respective `std` feature). Do double-check me on that though, I'm not 100% sure!

Robbepop replied:
The problem is that this requires not having `rand`, `byteorder` and `rustc-hex` being activated but their features `rand-support`, `byteorder-support` and `rustc-hex-support` should be enabled since the entire code is guarded by them and not by their respective features which is more accurate I think. So this would cause problems I think.

Robbepop replied:
Btw.: I think the normal use case is to either use `std` feature or `libc` but not both at the same time. So implying the one by the other seems to be wrong to me. But I might be confused about that!

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
+rand = { version = "0.5", optional = true, default-features = false }
+rustc-hex = { version = "2.0", optional = true, default-features = false }
+quickcheck = { version = "0.7", optional = true }
+byteorder = { version = "1.2.7", optional = true, default-features = false }
```
we can

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
-serde = { version = "1.0", optional = true }
-serde_derive = { version = "1.0", optional = true }
+rand = { version = "0.5", optional = true, default-features = false }
+rustc-hex = { version = "2.0", optional = true, default-features = false }
```
Well we could disable its feature guard by default.

dvdplm replied:
From what I've seen in other places `rustc-hex` is all over the place so doing that would be quite some (boring) work. Probably not something we have time for right now... :)

Robbepop commented on .travis.yml:
```diff
@@ -9,7 +9,7 @@ script:
   - cargo build --all
   - cargo test --all --exclude uint --exclude fixed-hash
-  - cd fixed-hash/ && cargo test --features=std,heapsizeof,uint_conversions && cd ..
+  - cd fixed-hash/ && cargo test --features=quickcheck-support,heapsize-support && cd ..
```
Good idea!

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
 
-		impl AsRef<$from> for $from {
+		impl $name {
+			/// Returns a new fixed hash where all bits are set.
```
Oh my this doc-comment is outdated. The constructor was `all_set` before but `repeat_byte` was a more general constructor and replaced it.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
 			#[inline]
-			fn as_ref(&self) -> &$from {
-				&self
+			pub fn repeat_byte(byte: u8) -> $name {
```
It replaces usage of `Hash::from(array)`, e.g. `Hash::from([0x42; 5])` or `Hash::from([0x1, 0x2, 0x3, ...])`.
However, it has the great advantage over `Hash::from([0x42; 5])` to be able to leave away the number of bytes and be correct; so it is a simpler API that should be favored in my opinion over `Hash::from([0x42; 5])`.

Robbepop replied:
Prefer `Hash::repeat_byte(0x42)` over `Hash::from([0x42; N])` where `N` is the number of bytes in `Hash`.

dvdplm replied:
How about `with_repeated_byte()`?

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
+			///
+			/// If the length of `src` and the number of bytes in `self` do not match.
+			pub fn assign_from_slice(&mut self, src: &[u8]) {
+				$crate::core::assert_eq!(src.len(), $n_bytes);
```
I disagree very strongly with this argumentation to be honest. It was already a bad idea to disable assertions in C/C++ context at that time. What we could do is provide some `unsafe` APIs that very explicitely document this behaviour and only assert things in debug builds so that they stay debuggable.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
+			///
+			/// If the length of `src` and the number of bytes in `Self` do not match.
+			pub fn from_slice(src: &[u8]) -> Self {
+				$crate::core::assert_eq!(src.len(), $n_bytes);
```
see above

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
-			)]
-			pub fn contains<'a>(&'a self, b: &'a Self) -> bool {
+			#[inline]
+			pub fn covers(&self, b: &Self) -> bool {
```
It is not new, it is just the renamed `contains` method. But `contains` was very confusion in the context of it being used outside of a container type. So `covers` actually is a better name for the same functionality. I don't know where it is used tbh.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
+		}
 
-			impl_std_for_hash_internals!($from, $size);
+		/// Utilizies using the `byteorder` crate.
```
yep, thanks! :)

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
 
-			impl_std_for_hash_internals!($from, $size);
+		/// Utilizies using the `byteorder` crate.
+		#[cfg(feature = "byteorder-support")]
```
The `*_le` and `*_ne` APIs have been added for symmetry purposes. We could remove them again if you dislike bigger APIs for the purpose of symmetry.

dvdplm replied:
I'm ok with:
- the name change `low_u64()` –> `to_low_u64_be()`
- using the `byterorder` vs rolling our own
- adding sibling methods `to_low_u64_*()`

I do not think there's a reason to add the `from_low_u64_*()`. If there's a need we can add them later.

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
+			///
+			/// If `n` is greater than the number of bytes in `self`.
+			#[inline]
+			fn least_significant_bytes(&self, n: usize) -> &[u8] {
```
Docs won't be shown in default mode. However, I like documenting private parts as well since you never know when they will be published. And besides that documentation doesn't hurt for maintenance purposes, right? :)

Robbepop commented on fixed-hash/src/hash.rs:
```diff
@@ -6,735 +6,646 @@
+			/// If `n` is greater than the number of bytes in `self`.
+			#[inline]
+			fn least_significant_bytes(&self, n: usize) -> &[u8] {
+				$crate::core::assert_eq!(true, n <= Self::len_bytes());
```
see above

Robbepop commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,438 @@
+	}
+}
+
+mod covers {
```
see above. `covers` is just the renamed version of the old and deprecated `contains` method.

Robbepop commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,438 @@
+
+	#[test]
+	fn complex_covers() {
+		#[rustfmt::skip]
```
should be fixed by now

Robbepop commented on fixed-hash/src/tests.rs:
```diff
@@ -0,0 +1,438 @@
+		0x46, 0xB3, 0x7D, 0x11, 0x06,
+	]);
+	let h256 = H256::from(h160);
+	let test = H256::from([
```
yeah may be a better name

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
+rustc-hex-support = ["rustc-hex"]
+rand-support = ["rand"]
+byteorder-support = ["byteorder"]
+heapsize-support = ["heapsize"]
```
Good point and good suggestion! Will do just that!

Robbepop replied:
Btw.: I don't know why or if heapsize really requires libc. Will have to find out that first.

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
+rand-support = ["rand"]
+byteorder-support = ["byteorder"]
+heapsize-support = ["heapsize"]
+quickcheck-support = ["quickcheck"]
```
I saw [this](https://github.com/paritytech/parity-common/issues/62) issue in `parity-common` and thought that `quickcheck` might still play a role in certain projects. I am not sure if dependencies of `fixed-hash` are using it though.

Robbepop commented on fixed-hash/Cargo.toml:
```diff
@@ -1,28 +1,35 @@
+heapsize-support = ["heapsize"]
+quickcheck-support = ["quickcheck"]
+
+api-dummy = []
```
Good idea!

Robbepop commented on fixed-hash/README.md:
````diff
@@ -1,30 +1,36 @@
 ```
 
-Add conversions between a hash type and the equivalently sized unsigned int:
+Opt-in to add conversions between differently sized hashes.
````
I can do that once this has been merged.

Robbepop commented on fixed-hash/src/lib.rs:
```diff
@@ -8,43 +8,50 @@
 pub extern crate rand;
 
-#[cfg(feature="impl_quickcheck_arbitrary")]
+#[cfg(feature = "quickcheck-support")]
```
So always thought it is used by crates from the outside but I am totally not sure about that.

Robbepop commented on fixed-hash/src/hash.rs:
````diff
@@ -6,735 +6,646 @@
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// # // extern crate serde;
+/// # // #[macro_use] extern crate serde_derive;
````
Good point!

Robbepop replied:
Ehm ... funny enough the lines
```
extern crate serde;
#[macro_use] extern crate serde_derive;
```
Are commented-out in this doc-test for a reason. They compile-error when they are not commented-out.
However, we are both pretty sure about the fact that we need them in reality - so what should we do?

Robbepop commented on fixed-hash/src/hash.rs:
````diff
@@ -6,735 +6,646 @@
+/// ```
+///
+/// As always, visibility modifier is optional.
+/// So for a private type go ahead with this.
````
Sounds a lot better, thanks!

dvdplm commented on fixed-hash/src/hash.rs:
```diff
@@ -506,11 +506,7 @@ macro_rules! construct_hash {
-			feature = "libc",
-			not(target_os = "unknown")
-		))]
+		#[cfg(feature = "heapsize-support")]
```
The `target_os` need to be there I'm afraid. WASM can't use the `heapsize` crate.

Robbepop replied:
okay will revert it

Robbepop replied:
done

Robbepop commented on fixed-hash/README.md:
```diff
@@ -39,3 +39,29 @@ construct_fixed_hash!{
+
+### Available Features
+
+- `std`: Use the standard library instead of the core library.
```
good idea!

Robbepop replied:
done

dvdplm commented:
LGTM. Great work and perseverence here @Robbepop! Many thanks.
